File: apps/hora/vendored/ncoe-contracts/venture_spec.py
"""
VentureSpec — Python dataclasses mirroring contracts/venture_spec.ts.

This is the NCOE side of the data contract with AEGIS. NCOE constructs
``VentureSpec`` instances and serialises them to JSON; AEGIS consumes the JSON
via the TypeScript types in venture_spec.ts.

SCOPE: This contract is for Substrate Mode only. Academy is a separate
AEGIS in-game mode with its own data model, its own curriculum, and its
own ECFL certification path. **Academy MUST stay pristine** — no field
in this schema references, indexes, or promotes content into Academy.
If you need to express anything Academy-related, you are in the wrong
data model.

Validation rules enforced here:
    - ContributionScoring weights must sum to 1.0
    - SpinoutPolicy.contributor_equity_pool must be in [0, 1)
    - difficulty must be 1..5

When updating: keep this file and venture_spec.ts in lockstep — same
field names, same enum values, same defaults where they appear.
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from typing import Any, Literal, Optional


# ── Enums (mirrored as Literals for typing) ──────────────────────

ToolkitKind = Literal[
    "business_automation",
    "ai_agent",
    "data_feed",
    "financial_product",
    "ui_panel",
    "substrate_briefing",
    "trading_algorithm",
]

VisibilityKind = Literal["public", "subscribers", "owner_only"]
SLALevel = Literal["best_effort", "subscribed", "premium"]
BriefingFormat = Literal[
    "5_min_read", "interactive_module", "video", "scenario"]
BriefingStatus = Literal["draft", "published", "featured", "archived"]
TargetBuyer = Literal["B2B", "B2C", "BOTH"]
PivotKind = Literal["pricing", "segment", "packaging", "gtm", "feature"]
AttemptStatus = Literal[
    "in_progress", "shipped", "abandoned", "failed", "exited"]


# ── InterfaceSpec ────────────────────────────────────────────────

@dataclass
class InterfaceSpec:
    visibility: VisibilityKind = "public"
    callable_by_other_code: bool = False
    input_schema: Optional[dict[str, Any]] = None
    output_schema: Optional[dict[str, Any]] = None
    refresh_seconds_min: Optional[float] = None
    sla: SLALevel = "best_effort"


# ── ToolkitPiece ─────────────────────────────────────────────────

@dataclass
class ToolkitPiece:
    kind: ToolkitKind
    purpose: str
    required: bool
    description: Optional[str] = None
    interface: Optional[InterfaceSpec] = None


# ── SubstrateBriefingSpec ────────────────────────────────────────

@dataclass
class SubstrateBriefingSpec:
    """Player + Athena co-generated educational module, scoped entirely
    to Substrate Mode.

    Briefings live exclusively in Substrate Mode. They do not, and will
    never, carry an Academy/ECFL eligibility flag — Academy is a separate
    pristine pipeline with its own governance. If you find yourself
    wanting to add an Academy-related field here, you are in the wrong
    data model.
    """
    title: str
    purpose: str
    estimated_minutes: int
    format: BriefingFormat
    athena_prompt_seed: Optional[str] = None


# ── Win conditions ───────────────────────────────────────────────

@dataclass
class WinPredicate:
    type: str
    value: float | str
    description: Optional[str] = None


@dataclass
class WinConditions:
    first_revenue_within_days: Optional[int] = None
    survive_days: Optional[int] = None
    profitability_ratio_target: Optional[float] = None
    exit_multiple_target: Optional[float] = None
    custom: list[WinPredicate] = field(default_factory=list)


# ── Telemetry capture flags ──────────────────────────────────────

@dataclass
class TelemetryCapture:
    athena_chat_logs: bool = True
    code_versions: bool = True
    pricing_decisions: bool = True
    customer_telemetry: bool = True
    pivot_events: bool = True
    ui_iteration_count: bool = True
    per_piece_metrics: bool = True
    generation_metadata: bool = True


# ── Contribution scoring (weights MUST sum to 1.0) ───────────────

@dataclass
class ContributionScoring:
    outcome_quality: float = 0.30
    originality: float = 0.25
    adoption_in_compile: float = 0.20
    peer_engagement: float = 0.15
    time_invested: float = 0.10

    def __post_init__(self):
        total = (self.outcome_quality + self.originality
                 + self.adoption_in_compile + self.peer_engagement
                 + self.time_invested)
        if abs(total - 1.0) > 1e-6:
            raise ValueError(
                f"ContributionScoring weights must sum to 1.0; got {total:.6f}. "
                f"Current weights: outcome={self.outcome_quality} "
                f"originality={self.originality} "
                f"adoption={self.adoption_in_compile} "
                f"peer={self.peer_engagement} "
                f"time={self.time_invested}"
            )


# ── Spinout policy ───────────────────────────────────────────────

@dataclass
class SpinoutPolicy:
    """Real-world spinout config.

    The default 10% pool reserved for top contributors matches the
    designed equity-on-spinout structure. Distribution to specific
    players happens at spinout time, not gameplay time, to keep the
    in-game promise out of unregistered-securities territory. See
    contracts/README.md and the ToS for the legal framing.
    """
    spinout_eligible: bool = True
    contributor_equity_pool: float = 0.10
    top_contributors_invited: int = 10
    jurisdiction: str = "PT"

    def __post_init__(self):
        if not 0.0 <= self.contributor_equity_pool < 1.0:
            raise ValueError(
                f"contributor_equity_pool must be in [0, 1); "
                f"got {self.contributor_equity_pool}"
            )
        if self.top_contributors_invited < 1:
            raise ValueError(
                f"top_contributors_invited must be >= 1; "
                f"got {self.top_contributors_invited}"
            )


# ── Source metadata ──────────────────────────────────────────────

@dataclass
class VentureSource:
    pipeline: str = "ncoe_multi_agent"
    pipeline_version: str = "0.1.0"
    domain: str = "optimization"
    seed: Optional[int] = None
    problem_hash: Optional[str] = None


# ── The top-level VentureSpec ────────────────────────────────────

@dataclass
class VentureSpec:
    venture_id: str
    pitch: str
    description: str
    market_summary: str
    starting_capital: float
    target_eta_mvp_days: int
    target_buyer: TargetBuyer
    difficulty: int
    implementation_pieces_required: list[ToolkitPiece]
    spec_version: str = "1.0"
    generated_at: str = field(default_factory=lambda:
        datetime.now(timezone.utc).isoformat())
    source: VentureSource = field(default_factory=VentureSource)
    top_objections: list[str] = field(default_factory=list)
    implementation_pieces_optional: list[ToolkitPiece] = field(default_factory=list)
    substrate_briefings: list[SubstrateBriefingSpec] = field(default_factory=list)
    win_conditions: WinConditions = field(default_factory=WinConditions)
    telemetry_capture: TelemetryCapture = field(default_factory=TelemetryCapture)
    contribution_scoring: ContributionScoring = field(default_factory=ContributionScoring)
    spinout_policy: SpinoutPolicy = field(default_factory=SpinoutPolicy)
    tags: list[str] = field(default_factory=list)

    def __post_init__(self):
        if not 1 <= self.difficulty <= 5:
            raise ValueError(
                f"difficulty must be in 1..5; got {self.difficulty}")
        if self.starting_capital < 0:
            raise ValueError(
                f"starting_capital must be >= 0; got {self.starting_capital}")
        if self.target_eta_mvp_days < 1:
            raise ValueError(
                f"target_eta_mvp_days must be >= 1; "
                f"got {self.target_eta_mvp_days}")
        if not self.implementation_pieces_required:
            raise ValueError(
                "VentureSpec must have at least one required toolkit piece")

    def to_json(self, **kwargs) -> str:
        """Serialise to JSON for sending across the NCOE→AEGIS boundary."""
        return json.dumps(asdict(self), **kwargs)


# ── Telemetry payloads (AEGIS → Quadratic) ───────────────────────

@dataclass
class ToolkitPieceAttempt:
    piece_kind: ToolkitKind
    artifact_hash: str
    version_count: int
    last_updated_at: str
    metrics: dict[str, float] = field(default_factory=dict)


@dataclass
class SubstrateBriefingAttempt:
    briefing_id: str
    title: str
    format: BriefingFormat
    rating_count: int = 0
    views: int = 0
    avg_rating: Optional[float] = None
    status: BriefingStatus = "draft"


@dataclass
class VentureMetrics:
    total_revenue: float = 0.0
    subscriber_count: int = 0
    first_revenue_at: Optional[str] = None
    distinct_buyers: int = 0
    customer_b2b_ratio: float = 0.0
    retention_30d: float = 0.0


@dataclass
class PivotEvent:
    at: str
    kind: PivotKind
    before: str
    after: str
    notes: Optional[str] = None


@dataclass
class VentureAttempt:
    attempt_id: str
    venture_id: str
    player_id: str
    company_id: str
    started_at: str
    status: AttemptStatus
    ended_at: Optional[str] = None
    pieces: list[ToolkitPieceAttempt] = field(default_factory=list)
    briefings: list[SubstrateBriefingAttempt] = field(default_factory=list)
    metrics: VentureMetrics = field(default_factory=VentureMetrics)
    pivots: list[PivotEvent] = field(default_factory=list)
    contribution_score: Optional[float] = None
    athena_session_ids: list[str] = field(default_factory=list)

File: apps/hora/vendored/ncoe-contracts/venture_spec.ts
/**
 * VentureSpec — Data contract between NCOE and AEGIS Empire.
 *
 * NCOE generates VentureSpec objects (one per startup idea). AEGIS reads them
 * and renders them as Venture Board cards inside Substrate Mode / Campaign.
 * AEGIS sends VentureAttempt telemetry back to Quadratic for compilation.
 *
 * SCOPE: This contract is for Substrate Mode only. Academy is a separate
 * AEGIS in-game mode with its own data model, its own curriculum, and its
 * own ECFL certification path — it does NOT cross this boundary. Content
 * generated inside Substrate Mode (Briefings, ventures, etc.) lives
 * entirely within Substrate Mode.
 *
 * The schema is the single source of truth — both Python (NCOE) and TypeScript
 * (AEGIS) sides MUST match. When updating, edit BOTH files in lockstep.
 */

// ── Toolkit categories ────────────────────────────────────────────

export type ToolkitKind =
  | "business_automation"   // workflow agent (e.g. fleet routing optimizer)
  | "ai_agent"              // an Athena variant tuned to a domain
  | "data_feed"             // curated/processed in-game data stream
  | "financial_product"     // contract: insurance, derivative, structured
  | "ui_panel"              // a custom panel/UI players see in-game
  | "substrate_briefing"    // short practical module — NOT ECFL-eligible
  | "trading_algorithm";    // automated trading logic

// (Academy / ECFL are intentionally NOT modelled here. They are pristine
// and live entirely outside this contract. This schema only describes
// Substrate Mode artefacts. There is no field in this contract that
// references, indexes, promotes to, or otherwise contaminates Academy
// content. If you find yourself reaching for one, you are in the wrong
// data model — Academy has its own.)


// ── Per-piece interface contract ──────────────────────────────────

export interface InterfaceSpec {
  /** Who can call/consume this piece. */
  visibility: "public" | "subscribers" | "owner_only";
  /** Whether other player-built code can invoke this piece. */
  callable_by_other_code: boolean;
  /** JSON Schema for input shape, when applicable. */
  input_schema?: object;
  /** JSON Schema for output shape, when applicable. */
  output_schema?: object;
  /** Min refresh / max latency, in seconds (data feeds, real-time pieces). */
  refresh_seconds_min?: number;
  /** Service-level expectation. */
  sla?: "best_effort" | "subscribed" | "premium";
}


// ── Implementation pieces ────────────────────────────────────────

export interface ToolkitPiece {
  /** Stable category identifier — see ToolkitKind. */
  kind: ToolkitKind;
  /** Short purpose phrase (e.g. "fleet routing optimizer"). */
  purpose: string;
  /** Whether the player MUST build this piece to ship the venture. */
  required: boolean;
  /** Player-facing explanation. */
  description?: string;
  /** Programmatic interface contract — only for code-bearing pieces. */
  interface?: InterfaceSpec;
}


// ── Substrate Briefings (player-generated educational content) ───

export interface SubstrateBriefingSpec {
  title: string;
  purpose: string;
  estimated_minutes: number;
  format: "5_min_read" | "interactive_module" | "video" | "scenario";
  /** Optional seed prompt Athena uses when co-creating with the player. */
  athena_prompt_seed?: string;
  // NOTE: Briefings live exclusively in Substrate Mode. They do not, and
  // will never, carry an Academy/ECFL eligibility flag. Academy content
  // is authored in a separate pipeline with its own governance.
}


// ── Win conditions / success criteria ────────────────────────────

export interface WinPredicate {
  /** Identifier the venture engine dispatches on. */
  type: string;
  /** Threshold/target — depends on `type`. */
  value: number | string;
  description?: string;
}

export interface WinConditions {
  /** Days until first in-game revenue counts as a "success". */
  first_revenue_within_days?: number;
  /** Days the venture must remain operational without abandoning. */
  survive_days?: number;
  /** revenue/cost ratio threshold for "profitable". */
  profitability_ratio_target?: number;
  /** Exit valuation as multiple of starting_capital. */
  exit_multiple_target?: number;
  /** Free-form additional predicates evaluated by the venture engine. */
  custom?: WinPredicate[];
}


// ── What gets logged for compilation later ───────────────────────

export interface TelemetryCapture {
  athena_chat_logs: boolean;
  code_versions: boolean;
  pricing_decisions: boolean;
  customer_telemetry: boolean;
  pivot_events: boolean;
  ui_iteration_count: boolean;
  per_piece_metrics: boolean;
  generation_metadata: boolean;
}


// ── Player contribution scoring (for spinout equity pool) ────────

/**
 * Weights MUST sum to 1.0. The Python side validates this at construction;
 * AEGIS should validate before posting to the Quadratic backend.
 */
export interface ContributionScoring {
  outcome_quality: number;
  originality: number;
  adoption_in_compile: number;
  peer_engagement: number;
  time_invested: number;
}


// ── Real-world spinout policy ────────────────────────────────────

export interface SpinoutPolicy {
  /** Whether this venture is eligible for real-world spinout at all. */
  spinout_eligible: boolean;
  /** Equity pool reserved for top contributors (0..1). E.g. 0.10 = 10%. */
  contributor_equity_pool: number;
  /** N players invited to the formal cap table on spinout. */
  top_contributors_invited: number;
  /**
   * Jurisdiction governing the spinout cap table.
   * Default "PT" (Portugal) given founder's incorporation choice.
   */
  jurisdiction: "PT" | "EE" | "US-DE" | "UK" | string;
}


// ── Source metadata (audit trail back to NCOE) ───────────────────

export interface VentureSource {
  pipeline: "ncoe_multi_agent";
  pipeline_version: string;
  domain: "optimization" | "software" | "engineering" | "chemistry"
        | "marketing" | "finance" | string;
  seed?: number;
  /** Hash of the input problem instance, if applicable. */
  problem_hash?: string;
}


// ── The top-level VentureSpec ────────────────────────────────────

export interface VentureSpec {
  venture_id: string;
  spec_version: "1.0";
  generated_at: string;     // ISO-8601 UTC
  source: VentureSource;

  // Player-facing summary
  pitch: string;
  description: string;
  market_summary: string;
  top_objections: string[];
  difficulty: 1 | 2 | 3 | 4 | 5;
  starting_capital: number;
  target_eta_mvp_days: number;
  target_buyer: "B2B" | "B2C" | "BOTH";

  // What needs to be built
  implementation_pieces_required: ToolkitPiece[];
  implementation_pieces_optional: ToolkitPiece[];
  substrate_briefings: SubstrateBriefingSpec[];

  // Engine config
  win_conditions: WinConditions;
  telemetry_capture: TelemetryCapture;
  contribution_scoring: ContributionScoring;
  spinout_policy: SpinoutPolicy;

  tags: string[];
}


// ── AEGIS → Quadratic telemetry payloads ─────────────────────────

export interface ToolkitPieceAttempt {
  piece_kind: ToolkitKind;
  /** Hash of the latest in-game code/config for this piece. */
  artifact_hash: string;
  /** Number of versions iterated by the player. */
  version_count: number;
  last_updated_at: string;
  /** Per-piece performance metrics — keys depend on `piece_kind`. */
  metrics: Record<string, number>;
}

export interface SubstrateBriefingAttempt {
  briefing_id: string;
  title: string;
  format: SubstrateBriefingSpec["format"];
  avg_rating?: number;     // 0..5
  rating_count: number;
  views: number;
  /** Lifecycle within Substrate Mode (not Academy). */
  status: "draft" | "published" | "featured" | "archived";
}

export interface VentureMetrics {
  total_revenue: number;
  subscriber_count: number;
  first_revenue_at?: string;
  /** Distinct buyer entities (companies + personal wallets, deduped). */
  distinct_buyers: number;
  /** Fraction of revenue from B2B subscribers (0..1). */
  customer_b2b_ratio: number;
  /** 30-day customer retention (0..1). */
  retention_30d: number;
}

export interface PivotEvent {
  at: string;
  kind: "pricing" | "segment" | "packaging" | "gtm" | "feature";
  before: string;
  after: string;
  notes?: string;
}

export interface VentureAttempt {
  attempt_id: string;
  venture_id: string;
  player_id: string;
  /** AEGIS company ID used to develop this venture. */
  company_id: string;
  started_at: string;
  ended_at?: string;
  status: "in_progress" | "shipped" | "abandoned" | "failed" | "exited";
  pieces: ToolkitPieceAttempt[];
  briefings: SubstrateBriefingAttempt[];
  metrics: VentureMetrics;
  pivots: PivotEvent[];
  /** Computed by Quadratic's contribution scorer (0..1). */
  contribution_score?: number;
  /** Athena conversation IDs used during co-creation (for spec extraction). */
  athena_session_ids: string[];
}
